### PR TITLE
[Fix #48688] Duplicate callback execution when child autosaves parent with `has_one` and `belongs_to`

### LINKED ISSUE #48688: Callbacks called twice when child accepts_nested_attributes_for parent
### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :name
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
    t.string :name
  end
end

class Post < ActiveRecord::Base
  has_one :comment

  before_validation :before_validation_callback_post
  after_validation :after_validation_callback_post

  before_save :before_save_callback_post
  after_save :after_save_callback_post

  before_create :before_create_callback_post
  after_create :after_create_callback_post

  def before_validation_callback_post
    puts 'Post: before_validation_callback'
  end

  def after_validation_callback_post
    puts 'Post: after_validation_callback'
  end

  def before_save_callback_post
    puts 'Post: before_save_callback'
  end

  def after_save_callback_post
    puts 'Post: after_save_callback'
  end

  def before_create_callback_post
    puts 'Post: before_create_callback'
  end

  def after_create_callback_post
    puts 'Post: after_create_callback'
  end
end

class Comment < ActiveRecord::Base
  belongs_to :post

  accepts_nested_attributes_for :post

  before_validation :before_validation_callback_comment
  after_validation :after_validation_callback_comment

  before_save :before_save_callback_comment
  after_save :after_save_callback_comment

  before_create :before_create_callback_comment
  after_create :after_create_callback_comment

  def before_validation_callback_comment
    puts 'Comment: before_validation_callback'
  end

  def after_validation_callback_comment
    puts 'Comment: after_validation_callback'
  end

  def before_save_callback_comment
    puts 'Comment: before_save_callback'
  end

  def after_save_callback_comment
    puts 'Comment: after_save_callback'
  end

  def before_create_callback_comment
    puts 'Comment: before_create_callback'
  end

  def after_create_callback_comment
    puts 'Comment: after_create_callback'
  end
end

class BugTest < Minitest::Test
  def test_association_stuff
    Comment.create!(name: 'foo', post_attributes: {name: 'bar'})
  end
end

```

### Expected behavior
`before_validation`, `after_validation`, `before_save`, `after_save` callbacks should only be called once when saving a parent record that accepts nested attributes for a child record. Note that `before_create`, `after_create` callbacks **are** only called once, so this inconsistency is especially confusing.

### Actual behavior
`before_validation`, `after_validation`, `before_save`, `after_save` are all called twice on parent record.

```
(base) m-apd-jmiller:models jmiller$ ruby active_record_main.rb 
Fetching https://github.com/rails/rails.git
Fetching gem metadata from https://rubygems.org/......
Resolving dependencies...
Using rake 13.0.6
Using connection_pool 2.4.1
Using minitest 5.18.1
Using webrick 1.8.1
Using erubi 1.12.0
Using rack 3.0.8
Using websocket-extensions 0.1.5
Using marcel 1.0.2
Using rack-test 2.1.0
Using bundler 2.4.7
Using websocket-driver 0.7.5
Using nio4r 2.5.9
Using zeitwerk 2.6.8
Using date 3.3.3
Using timeout 0.4.0
Using builder 3.2.4
Using io-console 0.6.0
Using racc 1.7.1
Using crass 1.0.6
Using thor 1.2.2
Using sqlite3 1.6.3 (x86_64-darwin)
Using rack-session 2.0.0
Using mini_mime 1.1.2
Using rackup 2.1.0
Using concurrent-ruby 1.2.2
Using net-protocol 0.2.1
Using reline 0.3.5
Using i18n 1.14.1
Using net-imap 0.3.6
Using nokogiri 1.15.3 (x86_64-darwin)
Using net-pop 0.1.2
Using net-smtp 0.3.3
Using loofah 2.21.3
Using irb 1.7.1
Using rails-html-sanitizer 1.6.0
Using tzinfo 2.0.6
Using mail 2.8.1
Using activesupport 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using activemodel 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using rails-dom-testing 2.1.1
Using activerecord 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using actionview 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using globalid 1.1.0
Using actionpack 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using activejob 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using actioncable 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using activestorage 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using actionmailer 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using actionmailbox 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using actiontext 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using railties 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
Using rails 7.1.0.alpha from https://github.com/rails/rails.git (at main@a5fc471)
-- create_table(:posts, {:force=>true})
D, [2023-07-07T12:17:20.483607 #2897] DEBUG -- :    (0.1ms)  DROP TABLE IF EXISTS "posts"
D, [2023-07-07T12:17:20.485077 #2897] DEBUG -- :    (1.1ms)  CREATE TABLE "posts" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "name" varchar)
   -> 0.0286s
-- create_table(:comments, {:force=>true})
D, [2023-07-07T12:17:20.485698 #2897] DEBUG -- :    (0.0ms)  DROP TABLE IF EXISTS "comments"
D, [2023-07-07T12:17:20.486042 #2897] DEBUG -- :    (0.1ms)  CREATE TABLE "comments" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "post_id" integer, "name" varchar)
   -> 0.0009s
D, [2023-07-07T12:17:20.488263 #2897] DEBUG -- :    (0.1ms)  CREATE TABLE "schema_migrations" ("version" varchar NOT NULL PRIMARY KEY)
D, [2023-07-07T12:17:20.490839 #2897] DEBUG -- :    (0.1ms)  CREATE TABLE "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
D, [2023-07-07T12:17:20.532633 #2897] DEBUG -- :   ActiveRecord::InternalMetadata Load (2.7ms)  SELECT * FROM "ar_internal_metadata" WHERE "ar_internal_metadata"."key" = ? ORDER BY "ar_internal_metadata"."key" ASC LIMIT 1  [[nil, "environment"]]
D, [2023-07-07T12:17:20.533215 #2897] DEBUG -- :   ActiveRecord::InternalMetadata Create (0.1ms)  INSERT INTO "ar_internal_metadata" ("key", "value", "created_at", "updated_at") VALUES ('environment', 'development', '2023-07-07 16:17:20.532735', '2023-07-07 16:17:20.532743')
Run options: --seed 61519

# Running:

Comment: before_validation_callback
Post: before_validation_callback
Post: after_validation_callback
Comment: after_validation_callback
Post: before_save_callback
Post: before_create_callback
D, [2023-07-07T12:17:20.661392 #2897] DEBUG -- :   TRANSACTION (0.1ms)  begin transaction
D, [2023-07-07T12:17:20.661656 #2897] DEBUG -- :   Post Create (0.4ms)  INSERT INTO "posts" ("name") VALUES (?)  [["name", "bar"]]
Comment: before_validation_callback
Comment: after_validation_callback
Comment: before_save_callback
Comment: before_create_callback
D, [2023-07-07T12:17:20.662908 #2897] DEBUG -- :   Comment Create (0.1ms)  INSERT INTO "comments" ("post_id", "name") VALUES (?, ?)  [["post_id", 1], ["name", "foo"]]
Comment: after_create_callback
Comment: after_save_callback
Post: after_create_callback
Post: after_save_callback
Comment: before_save_callback
Comment: after_save_callback
D, [2023-07-07T12:17:20.663528 #2897] DEBUG -- :   TRANSACTION (0.1ms)  commit transaction
```

### System configuration
**Rails version**: 5.2.8.1, 7.0.6 (tested both)

**Ruby version**: 2.7.6, 3.2.1 (tested both)

### PR DESCRIPTION
**Pull request by joshuay03**
### Motivation / Background

Fixes https://github.com/rails/rails/issues/48688

### Detail

This Pull Request updates `ActiveRecord::AutosaveAssociation#save_has_one_association` and `ActiveRecord::AutosaveAssociation#validate_single_association` to return early and not re-save the child record via autosave if the inverse association (`belongs_to`) has already set the child target, which happens after the child has been saved.

### Additional information

Some additional context in https://github.com/rails/rails/issues/48688#issuecomment-1774942159.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/test/cases/autosave_association_test.rb [hunk 1]
```diff
@@ -302,6 +302,14 @@ def test_callbacks_firing_order_on_save
     eye.update(iris_attributes: { color: "blue" })
     assert_equal [false, false, false, false], eye.after_save_callbacks_stack
   end
+
+  def test_callbacks_on_child_when_child_autosaves_parent
+    iris = Iris.create(eye: Eye.new)
+    assert_equal 1, iris.before_validation_callbacks_counter
+    assert_equal 1, iris.after_validation_callbacks_counter
+    assert_equal 1, iris.before_save_callbacks_counter
+    assert_equal 1, iris.after_save_callbacks_counter
```
Comment (joshuay03) [inline activerecord/test/cases/autosave_association_test.rb:311]: Before patch:
Screenshot 2023-10-30 at 6 21 11 pm

File: activerecord/test/cases/autosave_association_test.rb [hunk 2]
```diff
@@ -302,6 +302,22 @@ def test_callbacks_firing_order_on_save
     eye.update(iris_attributes: { color: "blue" })
     assert_equal [false, false, false, false], eye.after_save_callbacks_stack
   end
+
+  def test_callbacks_on_child_when_parent_autosaves_child
```
Comment (joshuay03) [inline activerecord/test/cases/autosave_association_test.rb:306]: Added this as a regression test.

File: activerecord/test/cases/autosave_association_test.rb [hunk 3]
```diff
@@ -302,6 +302,22 @@ def test_callbacks_firing_order_on_save
...
+    assert_equal 1, eye.iris.before_save_callbacks_counter
+    assert_equal 1, eye.iris.after_save_callbacks_counter
+  end
+
+  def test_callbacks_on_child_when_child_autosaves_parent
+    iris = Iris.create(eye: Eye.new)
+    assert_equal 1, iris.before_validation_callbacks_counter
+    assert_equal 1, iris.after_validation_callbacks_counter
+    assert_equal 1, iris.before_save_callbacks_counter
+    assert_equal 1, iris.after_save_callbacks_counter
```
Comment (joshuay03) [inline activerecord/test/cases/autosave_association_test.rb:319]: Before patch:
Screenshot 2023-10-30 at 6 21 11 pm

File: activerecord/test/cases/autosave_association_test.rb [hunk 4]
```diff
@@ -302,6 +302,26 @@ def test_callbacks_firing_order_on_save
...
+  end
+
+  def test_callbacks_on_child_when_child_autosaves_parent
+    iris = Iris.create(eye: Eye.new)
+    assert_equal 1, iris.before_validation_callbacks_counter
+    assert_equal 1, iris.before_create_callbacks_counter
+    assert_equal 1, iris.before_save_callbacks_counter
+    assert_equal 1, iris.after_validation_callbacks_counter
+    assert_equal 1, iris.after_create_callbacks_counter
+    assert_equal 1, iris.after_save_callbacks_counter
```
Comment (joshuay03) [inline activerecord/test/cases/autosave_association_test.rb:323]: Before patch:
Screenshot 2023-10-30 at 6 21 11 pm

**Note:** The `before_create` and `after_create` cases worked fine before the patch, I only added it in for consistency and to catch regressions.

**Review by cbisnett — COMMENTED** [review]
LGTM

I couldn't come up with any other edge cases that weren't already covered here.

**Comment by julian-berbel** [comment]
Is something blocking this? Recently ran into this issue

**Comment by joshuay03** [comment]
> Is something blocking this? Recently ran into this issue

I think I asked for a review a couple of times in the Discord server when I first opened this and then forgot about it. I'll rebase and ask again.

File: activerecord/lib/active_record/autosave_association.rb [hunk 5]
```diff
@@ -429,36 +434,44 @@ def save_collection_association(reflection)
...
             end
+
+            saved = record.save(validate: !autosave)
+            raise ActiveRecord::Rollback if !saved && autosave
+            saved
           end
         end
       end
 
+      def inverse_belongs_to_association_for(reflection, record)
```
Comment (casperisfine) [inline activerecord/lib/active_record/autosave_association.rb:469]: This isn't meant as public API right? So It would be best to move it in the private section

**Review by casperisfine — COMMENTED** [review]
One minor issue, other than that LGTM. The logic is hard to follow, the the tests are 💯

File: activerecord/lib/active_record/autosave_association.rb (hunk 5 above)
Comment (joshuay03) [reply activerecord/lib/active_record/autosave_association.rb:469]: I think it's already private. All these methods are under [this access modifier](https://github.com/rails/rails/blob/78ce2994b23b4b68db07a910df252284f1d18954/activerecord/lib/active_record/autosave_association.rb#L277)

Comment (casperisfine) [reply activerecord/lib/active_record/autosave_association.rb:469]: Ah indeed. I got confused because the methods around are documented, which is mostly for public methods.

**Comment by ghiculescu** [comment]
@joshuay03 I think this is causing https://github.com/rails/rails/issues/52304

When I ran the replication script against https://github.com/rails/rails/commit/cacf96ecd053f0a3d64c7bf478dac63f98e6aa6d, the last commit on main before this one, it passed.

**Comment by joshuay03** [comment]
> @joshuay03 I think this is causing #52304
> 
> When I ran the replication script against [cacf96e](https://github.com/rails/rails/commit/cacf96ecd053f0a3d64c7bf478dac63f98e6aa6d), the last commit on main before this one, it passed.

Thanks for the heads up, I'll have a look

**Comment by mjobin-mdsol** [comment]
will this be backported to 7.1 and 7.0 ? 

is 6.1 also affected?

**Comment by ghiculescu** [comment]
7.1 gets bug fixes, so it can be backported there. Other versions don't. See https://guides.rubyonrails.org/maintenance_policy.html

@joshuay03 is this fine to backport to 7.1?

**Comment by joshuay03** [comment]
> @joshuay03 is this fine to backport to 7.1?

I think so. I'm still a bit [concerned about other uncaught changes in behaviour](https://github.com/rails/rails/issues/52332#issuecomment-2227642793), but if it's okay to forward fix + backport if any come up, it's a 👍🏽 from me.

The backport will also need to include https://github.com/rails/rails/pull/52333 and https://github.com/rails/rails/issues/52332

**Comment by casperisfine** [comment]
> I'm still a bit [concerned about other uncaught changes in behaviour](https://github.com/rails/rails/issues/52332#issuecomment-2227642793),

Yeah me too, I know the GitHub folks are still having some trouble with this change, and on the Shopify side I think we're good now but still need to go past this upgrade. All this tells me that while this can be considered a bug fix, the behavior change is big enough that I don't think it's a good idea to backport it.
